sympy/matrices/determinant.py

```diff
@@ -409,10 +414,17 @@ def _charpoly(M, x='lambda', simplify=_simplify):
             m = m * (x - simplify(i))
         return PurePoly(m, x)
 
-    berk_vector = _berkowitz_vector(M)
-    x = uniquely_named_symbol(x, berk_vector, modify=lambda s: '_' + s)
-
-    return PurePoly([simplify(a) for a in berk_vector], x)
+    DOM = DomainMatrix.from_Matrix(M, extension=True)
+    domain = DOM.domain
+    if domain == 'EX' or isinstance(domain, FractionField):
+        if isinstance(domain, FractionField):
+            if all(isinstance(t, Symbol) for t in domain.symbols):
+                return _charpoly_DOM(DOM, M, x, domain)
+        berk_vector = _berkowitz_vector(M)
+        x = uniquely_named_symbol(x, berk_vector, modify=lambda s: '_' + s)
+        return PurePoly([simplify(a) for a in berk_vector], x)
```

Comment:
I think this could be better structured like:
```
if domain == EX:
    # Add a comment here explaining why
    use_domain = False
elif domain.is_FractionField and not all(t.is_Symbol for t in domain.symbols):
    # Add a comment here explaining why
    use_domain = False
else:
    # Add a comment indicating which domains would potentially be used here
    # e.g. ZZ, QQ, ZZ[x, y], QQ(x, y), QQ<sqrt(2)> etc
    use_domain = True

if use_domain:
    ...
else:
    ...
```
I don't think the domain will be a string so comparing with 'EX' does not work. Have you checked this?

Reply:
Yes, it was not looking good at all.
Thank you for the suggestion, will keep them in mind.

> I don't think the domain will be a string so comparing with 'EX' does not work. Have you checked this?
>
YES, I have removed the string, it was not working.

Thanks.

Reply:
Actually I guess that when the domain is a FractionField with non-symbol generators that doesn't mean that we would not want to use DomainMatrix. It just means that we would want to simplify the end result. Perhaps in that case we could get the result as a Poly and then do something like:
```
if simplify_poly:
    p = Poly(simplify(p.as_expr()), symbol)
```